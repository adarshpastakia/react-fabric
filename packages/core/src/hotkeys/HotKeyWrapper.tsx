/*
 * React Fabric
 * @version: 1.0.0
 *
 *
 * The MIT License (MIT)
 * Copyright (c) 2024 Adarsh Pastakia
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software
 * and associated documentation files (the "Software"), to deal in the Software without restriction,
 * including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense,
 * and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all copies or substantial
 * portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED
 * TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
 * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
 * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import {
  createContext,
  memo,
  useCallback,
  useContext,
  useRef,
  type FC,
  type PropsWithChildren,
} from "react";
import { useEffectDebugger } from "../hooks/useEffectDebugger";
import { reduceHotKey } from "./commons";

/** @internal */
export const HotKeysContext = createContext<{
  addHotKey: (key: string, handler: () => void, global?: boolean) => void;
  removeHotKey: (key: string) => void;
}>({} as AnyObject);

export const useHotKeys = () => useContext(HotKeysContext);

export const HotKeyWrapper: FC<PropsWithChildren> = memo(
  ({ children }: PropsWithChildren) => {
    const refEl = useRef<HTMLDivElement>(null);
    const keyList = useRef<KeyValue[]>([]);

    const addHotKey = useCallback(
      (keyCombo: string, handler: () => void, global = false) =>
        (keyList.current = [
          ...keyList.current.filter(({ _key }) => _key !== keyCombo),
          reduceHotKey({ keyCombo, handler, global }),
        ]),
      [],
    );
    const removeHotKey = useCallback(
      (key: string) =>
        (keyList.current = [
          ...keyList.current.filter(({ _key }) => _key !== key),
        ]),
      [],
    );

    const handler = useCallback((items: KeyValue[], event: KeyboardEvent) => {
      if (items.length === 0) return;
      const { key: keyCode, code, altKey, ctrlKey, metaKey, shiftKey } = event;
      if (
        keyCode !== "Enter" &&
        keyCode !== "Esc" &&
        ["INPUT", "TEXTAREA", "SELECT"].includes(
          (event.target as HTMLElement).tagName,
        )
      )
        return;
      const find = items.find(
        ({ key, alt, ctrl, meta, shift }: KeyValue) =>
          (key.toLowerCase() === keyCode.toLowerCase() ||
            key.toLowerCase() === code.toLowerCase()) &&
          alt === altKey &&
          ctrl === ctrlKey &&
          meta === metaKey &&
          shift === shiftKey,
      );
      if (find?.handler) {
        find.handler();
        event.preventDefault();
        event.stopImmediatePropagation();
      }
    }, []);

    const handleGlobal = useCallback(
      (event: KeyboardEvent) => {
        handler(
          keyList.current.filter((i) => i.global),
          event,
        );
      },
      [handler],
    );
    const handleHotKey = useCallback(
      (event: KeyboardEvent) => {
        handler(
          keyList.current.filter((i) => !i.global),
          event,
        );
      },
      [handler],
    );

    useEffectDebugger(
      () => {
        const el: HTMLElement | null | undefined = refEl.current?.parentElement;
        if (el) {
          el.tabIndex = 0;
          el.addEventListener("keydown", handleHotKey);
          document.addEventListener("keydown", handleGlobal);
          return () => {
            el.removeEventListener("keydown", handleHotKey);
            document.removeEventListener("keydown", handleGlobal);
          };
        }
      },
      [handleGlobal, handleHotKey],
      "HotKeyWrapper attach handlers",
    );

    return (
      <HotKeysContext.Provider value={{ addHotKey, removeHotKey }}>
        <div ref={refEl} style={{ display: "contents" }}>
          {children}
        </div>
      </HotKeysContext.Provider>
    );
  },
);
HotKeyWrapper.displayName = "HotKeyProvider";
