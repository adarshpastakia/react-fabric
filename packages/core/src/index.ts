/*
 * React Fabric
 * @version: 1.0.0
 *
 *
 * The MIT License (MIT)
 * Copyright (c) 2024 Adarsh Pastakia
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software
 * and associated documentation files (the "Software"), to deal in the Software without restriction,
 * including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense,
 * and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all copies or substantial
 * portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED
 * TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
 * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
 * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

export * from "./components";
export * from "./core";
export { addTranslationBundle } from "./i18n";
export { Flyout, Modal, Tooltip } from "./overlays";
export * from "./typography";

export * from "./hooks/useClientService";
export * from "./hooks/useDebounce";
export * from "./hooks/useEffectDebugger";
export * from "./hooks/useIsDark";
export * from "./hooks/useIsRtl";
export * from "./hooks/useNavigator";
export * from "./hooks/useOverlayService";
export * from "./hooks/useResize";
export * from "./hooks/useResizeObserver";
export * from "./hooks/useStorage";

export { HotKey } from "./hotkeys/HotKey";
export { HotKeyLabel } from "./hotkeys/HotKeyLabel";

export { CoreIcons } from "./types/icons";

export {
  ApplicationProvider,
  useApplicationContext,
  useGlobals,
  useNotificationService,
} from "./context/Global";
