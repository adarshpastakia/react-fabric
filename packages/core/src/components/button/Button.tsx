/*
 * React Fabric
 * @version: 1.0.0
 *
 *
 * The MIT License (MIT)
 * Copyright (c) 2024 Adarsh Pastakia
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software
 * and associated documentation files (the "Software"), to deal in the Software without restriction,
 * including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense,
 * and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all copies or substantial
 * portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED
 * TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
 * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
 * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import { isNil, isString, mergeRefs } from "@react-fabric/utilities";
import classNames from "classnames";
import {
  Fragment,
  isValidElement,
  type MouseEventHandler,
  useCallback,
  useEffect,
  useMemo,
  useRef,
  useState,
  type WheelEventHandler,
} from "react";
import { useEffectDebugger } from "../../hooks/useEffectDebugger";
import { HotKey } from "../../hotkeys/HotKey";
import { HotKeyLabel } from "../../hotkeys/HotKeyLabel";
import { Tooltip } from "../../overlays/tooltip/Tooltip";
import {
  type AriaProps,
  type BadgeType,
  type CallbackReturn,
  type ColorType,
  type CssProp,
  type IconProps,
  type PolymorphicProps,
  type RefProp,
  type SizeType,
  type TestProps,
  type TooltipType,
} from "../../types";
import { AnimationIndicator } from "../animations/Animations";
import { Badge, getBadgeProps } from "../badge/Badge";
import { Icon } from "../icon/Icon";
import classes from "./Button.module.css";

export interface BaseProps
  extends CssProp,
    AriaProps,
    IconProps,
    RefProp,
    TestProps {
  /**
   * primary icon alignment
   */
  iconAlign?: "start" | "end";
  /**
   * alternate icon align end (caret, arrow)
   */
  altIcon?: string;
  /**
   * button size
   */
  size?: SizeType;
  /**
   * button color
   */
  color?: ColorType;
  /**
   * button type
   */
  type?: "button" | "reset" | "submit";
  /**
   * button styling
   */
  variant?: "outline" | "solid" | "link";
  /**
   * active state with click disabled
   */
  active?: boolean;
  /**
   * loading state with animation and click disabled
   */
  loading?: boolean;
  /**
   * full rounded edges
   */
  rounded?: boolean;
  /**
   * spin icon on hover
   */
  spinOnHover?: boolean;
  /**
   * disabled state
   */
  disabled?: boolean;
  /**
   * stretch to full width
   */
  fullWidth?: boolean;
  /**
   * action completed message works best with `onClick` Promise return.
   * To hide completed message onClick/Promise return should be false
   */
  actionMessage?: string;
  /**
   * manually force show action message
   */
  showActionDone?: boolean;
  /**
   * show action message `onClick` or manually by setting `showActionDone` true
   */
  showActionDoneEvent?: "click" | "manual";
  /**
   * click handler, return Promise to autoset loading state
   */
  onClick?: (e?: React.MouseEvent) => CallbackReturn;
  /**
   * badges
   */
  badge?: string | number | BadgeType;
  /**
   * keyboard shortcut
   */
  hotKey?: string;

  onMouseOut?: MouseEventHandler;
  onMouseOver?: MouseEventHandler;
  onMouseMove?: MouseEventHandler;
  onMouseEnter?: MouseEventHandler;
  onMouseLeave?: MouseEventHandler;

  onWheel?: WheelEventHandler;
}

export type ButtonProps =
  | { children: string | React.ReactElement; "aria-label"?: string }
  | { children?: never; "aria-label": string };

/**
 * Clickable action button
 */
export const Button = <Tag extends React.ElementType = "button">({
  as,
  ref,
  children,
  className,
  active,
  color = "primary",
  icon,
  iconBg,
  iconColor,
  iconAlign = "start",
  altIcon,
  rtlFlip,
  rounded,
  size,
  disabled,
  loading,
  variant,
  badge,
  fullWidth,
  spinOnHover,
  onClick,
  actionMessage,
  type = "button",
  hotKey,
  showActionDone,
  showActionDoneEvent = "click",
  tabIndex = 0,
  // @ts-expect-error ignore
  "data-ref": dataRef,
  ...aria
}: BaseProps & ButtonProps & PolymorphicProps<Tag>) => {
  const refEl = useRef<AnyObject>(null);
  const [busy, setBusy] = useState(false);
  const [actionDone, setActionDone] = useState(false);

  const clickHandler = useCallback(
    (e: React.MouseEvent) => {
      setBusy(true);
      // e.stopPropagation();
      const ret = onClick?.(e);
      void Promise.resolve(ret).then((b) => {
        setBusy(false);
        if (b !== false && actionMessage && showActionDoneEvent === "click") {
          setActionDone(true);
        }
      });
    },
    [onClick, disabled, actionMessage, showActionDoneEvent],
  );

  /** ***************** reset done state on timeout *******************/
  useEffect(() => {
    actionMessage && showActionDone && setActionDone(true);
  }, [showActionDone, actionMessage]);
  useEffectDebugger(
    () => {
      if (actionDone) {
        const timer = setTimeout(() => setActionDone(false), 2000);
        return () => clearTimeout(timer);
      }
    },
    [actionDone],
    "ActionButton done handler",
  );

  const TooltipWrapper = useMemo(
    () => (actionMessage ? Tooltip : Fragment),
    [actionMessage, actionDone],
  );
  const tooltipProps = useMemo(() => {
    if (actionDone && actionMessage) {
      return {
        content: actionMessage,
        placement: "top",
        color,
        open: true,
      } as TooltipType;
    }
  }, [actionDone, actionMessage, color]);

  const badgeProps = useMemo(() => {
    return getBadgeProps(badge);
  }, [badge]);

  const E = as ?? "button";
  return (
    <div
      className={classNames(
        classes.buttonWrapper,
        className,
        "relative align-text-top select-none focus-within:z-1",
        fullWidth ? "block w-full" : "inline-block",
      )}
      data-ref={dataRef ?? "button"}
      data-size={size}
      data-loading={busy || loading}
      data-disabled={disabled}
    >
      {hotKey && (
        <HotKey
          keyCombo={hotKey}
          handler={() => !disabled && refEl.current?.click()}
        />
      )}
      <TooltipWrapper {...tooltipProps}>
        <E
          ref={mergeRefs(ref, refEl)}
          role="button"
          className={classNames(
            classes.button,
            "pointer-events-auto cursor-pointer select-none flex flex-nowrap w-full h-full items-center border-0 -outline-offset-1 justify-center max-w-72 p-0 font-sans font-medium appearance-none",
            `focus-visible:ring-2 focus-visible:ring-offset-2 buttonInner`,
            fullWidth && "max-w-full w-full",
            rounded && "rounded-full",
            spinOnHover && classes.spinOnHover,
          )}
          type={type}
          disabled={disabled}
          tabIndex={tabIndex}
          data-inner-clickable
          onClick={clickHandler}
          data-color={color}
          data-active={active}
          data-variant={variant}
          {...aria}
        >
          {icon && (
            <Icon
              className={classes.buttonIcon}
              icon={icon}
              bg={iconBg}
              color={iconColor}
              rtlFlip={rtlFlip}
              data-align={iconAlign}
            />
          )}
          {isValidElement(children) && (
            <div className={classes.buttonChild}>{children}</div>
          )}
          {isString(children) && (
            <label
              className={classNames(classes.buttonLabel, "truncate")}
              data-colored-icon={!!iconBg}
            >
              {children}
            </label>
          )}
          {!isNil(badge) && <Badge {...badgeProps} />}
          {hotKey && <HotKeyLabel keyCombo={hotKey} />}
          {altIcon && (
            <Icon className={classes.altIcon} icon={altIcon} rtlFlip />
          )}
        </E>
      </TooltipWrapper>
      {actionDone && (
        <div
          className={classNames(
            "absolute inset-0 overflow-hidden flex justify-center items-center",
            `bg-${color}-300`,
            rounded ? "rounded-full" : "rounded",
          )}
        >
          <AnimationIndicator type="check" color={color} />
        </div>
      )}
    </div>
  );
};
