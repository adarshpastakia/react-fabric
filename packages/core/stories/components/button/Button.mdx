import {
  Canvas,
  Description,
  Primary,
  Story,
  Title,
} from "@storybook/addon-docs";
import {
  Anchor,
  ArgTypes,
  Controls,
  Heading,
  Meta,
  Subheading,
} from "@storybook/blocks";
import { Fragment } from "react";
import { Button, Collapsable } from "../../../src";
import * as ButtonStories from "./Button.stories";
import * as ConfirmButtonStories from "./ConfirmButton.stories";

<Meta of={ButtonStories} />

<Title>Button</Title>

<Anchor storyId="button">
  <div className="hidden-anchor">
    <Subheading>Button</Subheading>
  </div>
</Anchor>

<Description of={Button} />

<Primary />

<Collapsable className="control-panel">
  <div>Controls</div>
  <Controls of={ButtonStories._Button} />
</Collapsable>

<Anchor storyId="styles">
  <Subheading>Styles</Subheading>
</Anchor>
<Description of={ButtonStories.Styles} />
<Canvas withToolbar of={ButtonStories.Styles} />
<Collapsable className="control-panel">
  <div>Controls</div>
  <Controls of={ButtonStories.Styles} />
</Collapsable>

<Anchor storyId="colors">
  <Subheading>Colors</Subheading>
</Anchor>
<Description of={ButtonStories.Colors} />
<Canvas withToolbar of={ButtonStories.Colors} />
<Collapsable className="control-panel">
  <div>Controls</div>
  <Controls of={ButtonStories.Colors} />
</Collapsable>

<Anchor storyId="sizes">
  <Subheading>Sizes</Subheading>
</Anchor>
<Description of={ButtonStories.Sizes} />
<Canvas withToolbar of={ButtonStories.Sizes} />
<Collapsable className="control-panel">
  <div>Controls</div>
  <Controls of={ButtonStories.Sizes} />
</Collapsable>

<Anchor storyId="states">
  <Subheading>States</Subheading>
</Anchor>
<Description of={ButtonStories.States} />
<Canvas withToolbar of={ButtonStories.States} />
<Collapsable className="control-panel">
  <div>Controls</div>
  <Controls of={ButtonStories.States} />
</Collapsable>

<Anchor storyId="badges">
  <Subheading>Badges</Subheading>
</Anchor>
<Description of={ButtonStories.Badges} />
<Canvas withToolbar of={ButtonStories.Badges} />
<Collapsable className="control-panel">
  <div>Controls</div>
  <Controls of={ButtonStories.Badges} />
</Collapsable>

<Anchor storyId="actionState">
  <Subheading>Action State</Subheading>
</Anchor>
<Description of={ButtonStories.ActionState} />
<Canvas withToolbar of={ButtonStories.ActionState} />
<Collapsable className="control-panel">
  <div>Controls</div>
  <Controls of={ButtonStories.ActionState} />
</Collapsable>

<hr />

<Anchor storyId="props">
  <Subheading>Props</Subheading>
</Anchor>

<ArgTypes sort="requiredFirst" />

<Anchor storyId="sample">
  <Subheading>Example</Subheading>
</Anchor>

```tsx
import { Button } from "@react-fabric/core";

export const Example = () => {
  return (
    <Button
      icon="icon svg path"
      altIcon="caret icon"
      color="color"
      variant="style"
      loading={boolean}
      disabled={boolean}
    >
      Button Label
    </Button>
  );
};
```
