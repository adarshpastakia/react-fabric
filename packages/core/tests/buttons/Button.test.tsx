/*
 * React Fabric
 * @version: 1.0.0
 *
 *
 * The MIT License (MIT)
 * Copyright (c) 2024 Adarsh Pastakia
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software
 * and associated documentation files (the "Software"), to deal in the Software without restriction,
 * including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense,
 * and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all copies or substantial
 * portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED
 * TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
 * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
 * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import { composeStories } from "@storybook/react";
import { fireEvent, render, waitFor } from "@testing-library/react";
import { CoreIcons } from "../../src";
import { HotKeyWrapper } from "../../src/hotkeys/HotKeyWrapper";
import * as stories from "../../stories/components/button/Button.stories";

const { Tester } = composeStories(stories);

describe("Button", () => {
  it("should render label only", () => {
    const fragment = render(<Tester>Click Me</Tester>);
    expect(
      fragment.container.querySelector("[data-ref='button'] > button"),
    ).not.toBeNull();
    expect(
      fragment.container.querySelector("[data-ref='button'] > button > dfn"),
    ).toBeNull();
    expect(
      fragment.container.querySelector("[data-ref='button'] > button > label")
        ?.innerHTML,
    ).toBe("Click Me");
    expect(document.body.innerHTML).toMatchSnapshot();
    fragment.unmount();
  });

  it("should render icon only", () => {
    const fragment = render(<Tester icon="mdi mdi-bell" />);
    expect(
      fragment.container.querySelector("[data-ref='button'] > button"),
    ).not.toBeNull();
    expect(
      fragment.container.querySelector("[data-ref='button'] > button > dfn"),
    ).not.toBeNull();
    expect(
      fragment.container.querySelector("[data-ref='button'] > button > label"),
    ).toBeNull();
    expect(document.body.innerHTML).toMatchSnapshot();
    fragment.unmount();
  });

  it("should render alt icon", () => {
    const fragment = render(
      <Tester
        icon="mdi mdi-bell"
        spinOnHover
        altIcon={CoreIcons.chevronDown}
      />,
    );
    expect(
      fragment.container.querySelector("[data-ref='button'] > button"),
    ).not.toBeNull();
    expect(
      fragment.container.querySelectorAll("[data-ref='button'] > button > dfn"),
    ).toHaveLength(2);
    expect(document.body.innerHTML).toMatchSnapshot();
    fragment.unmount();
  });

  it("should render rounded full-width", () => {
    const fragment = render(
      <Tester fullWidth rounded>
        Click Me!
      </Tester>,
    );
    expect(
      fragment.container.querySelector("[data-ref='button'] > button"),
    ).not.toBeNull();
    expect(document.body.innerHTML).toMatchSnapshot();
    fragment.unmount();
  });

  it("should handle click", () => {
    const fn = jest.fn();
    const fragment = render(<Tester onClick={fn} />);
    fireEvent.click(
      fragment.container.querySelector(
        "[data-ref='button'] > button",
      ) as HTMLElement,
    );
    expect(fn).toHaveBeenCalled();
    fragment.unmount();
  });

  it("should not handle click when disabled", () => {
    const fn = jest.fn();
    const fragment = render(<Tester disabled onClick={fn} />);
    fireEvent.click(
      fragment.container.querySelector(
        "[data-ref='button'] > button",
      ) as HTMLElement,
    );
    expect(fn).not.toHaveBeenCalled();
    fragment.unmount();
  });

  it("should handle hotkey", () => {
    const fn = jest.fn();
    const fragment = render(
      <HotKeyWrapper>
        <Tester onClick={fn} hotKey="shift+a" />
      </HotKeyWrapper>,
    );
    expect(
      fragment.container.querySelector<HTMLElement>(
        "[data-ref='button'] > button > kbd > span",
      )?.innerHTML,
    ).toBe("SHIFT A");
    fireEvent.keyDown(
      fragment.container.querySelector(
        "[data-ref='button'] > button",
      ) as HTMLElement,
      {
        key: "A",
        code: "KeyA",
        shiftKey: true,
        keyCode: 65,
        charCode: 0,
      },
    );
    expect(fn).toHaveBeenCalled();
    fragment.unmount();
  });

  it("should render with badge", () => {
    const fragment = render(<Tester badge={99}>Click Me</Tester>);
    expect(
      fragment.container.querySelector<HTMLElement>(
        "[data-ref='button'] > button [data-ref='badge'] > span",
      )?.innerHTML,
    ).toBe("99");
    expect(document.body.innerHTML).toMatchSnapshot();
    fragment.unmount();
  });

  it("should render action message", async () => {
    const fn = jest.fn(() => true);
    const fragment = render(
      <Tester onClick={fn} actionMessage="Action completed">
        Click Me
      </Tester>,
    );
    fireEvent.click(
      fragment.container.querySelector(
        "[data-ref='button'] > button",
      ) as HTMLElement,
    );
    await waitFor(() => {
      expect(
        fragment.container.querySelector("[data-ref='animated-indicator']"),
      ).not.toBeNull();
      expect(
        fragment.baseElement.querySelector("[data-ref='tooltip']"),
      ).not.toBeNull();
    });
    jest.runAllTimers();
    expect(document.body.innerHTML).toMatchSnapshot();
    fragment.unmount();
  });

  it("should render action message via prop", async () => {
    const fragment = render(
      <Tester
        actionMessage="Action completed"
        showActionDoneEvent="manual"
        showActionDone
        rounded
      >
        Click Me
      </Tester>,
    );
    await waitFor(() => {
      expect(
        fragment.container.querySelector("[data-ref='animated-indicator']"),
      ).not.toBeNull();
      expect(
        fragment.baseElement.querySelector("[data-ref='tooltip']"),
      ).not.toBeNull();
    });
    jest.runAllTimers();
    expect(document.body.innerHTML).toMatchSnapshot();
    fragment.unmount();
  });
});
