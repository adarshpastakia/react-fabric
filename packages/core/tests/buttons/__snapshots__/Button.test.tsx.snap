// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should render action message 1`] = `"<div></div><div></div><div></div><div></div><div></div><div></div><div tabindex="0"></div><div></div><div><div class="flex gap-2 items-center p-4"><div class="buttonWrapper relative align-text-top select-none focus-within:z-1 inline-block" data-ref="button"><button role="button" class="button pointer-events-auto cursor-pointer select-none flex flex-nowrap w-full h-full items-center border-0 -outline-offset-1 justify-center max-w-72 p-0 font-sans font-medium appearance-none focus-visible:ring-2 focus-visible:ring-offset-2 buttonInner" type="button" tabindex="0" data-inner-clickable="true" data-color="primary"><label class="buttonLabel truncate" data-colored-icon="false">Click Me</label></button></div></div></div>"`;

exports[`Button should render action message via prop 1`] = `"<div></div><div></div><div></div><div></div><div></div><div></div><div tabindex="0"></div><div></div><div></div><div><div class="flex gap-2 items-center p-4"><div class="buttonWrapper relative align-text-top select-none focus-within:z-1 inline-block" data-ref="button"><button role="button" class="button pointer-events-auto cursor-pointer select-none flex flex-nowrap w-full h-full items-center border-0 -outline-offset-1 justify-center max-w-72 p-0 font-sans font-medium appearance-none focus-visible:ring-2 focus-visible:ring-offset-2 buttonInner rounded-full" type="button" tabindex="0" data-inner-clickable="true" data-color="primary"><label class="buttonLabel truncate" data-colored-icon="false">Click Me</label></button></div></div></div>"`;

exports[`Button should render alt icon 1`] = `"<div></div><div></div><div><div class="flex gap-2 items-center p-4"><div class="buttonWrapper relative align-text-top select-none focus-within:z-1 inline-block" data-ref="button"><button role="button" class="button pointer-events-auto cursor-pointer select-none flex flex-nowrap w-full h-full items-center border-0 -outline-offset-1 justify-center max-w-72 p-0 font-sans font-medium appearance-none focus-visible:ring-2 focus-visible:ring-offset-2 buttonInner spinOnHover" type="button" tabindex="0" data-inner-clickable="true" data-color="primary"><dfn data-align="start" role="presentation" class="icon buttonIcon select-none overflow-hidden inline-block box-content leading-none not-italic text-center rounded" data-clickable="false"><i class="mdi mdi-bell"></i></dfn><dfn role="presentation" class="icon altIcon select-none overflow-hidden inline-block box-content leading-none not-italic text-center rounded" data-flip="true" data-clickable="false"><svg viewBox="0 0 24 24"><path fill="currentColor" d="M6.173 7.29l5.827 5.827 5.827-5.827 1.79 1.803-7.617 7.617-7.617-7.617 1.79-1.803z"></path></svg></dfn></button></div></div></div>"`;

exports[`Button should render icon only 1`] = `"<div></div><div><div class="flex gap-2 items-center p-4"><div class="buttonWrapper relative align-text-top select-none focus-within:z-1 inline-block" data-ref="button"><button role="button" class="button pointer-events-auto cursor-pointer select-none flex flex-nowrap w-full h-full items-center border-0 -outline-offset-1 justify-center max-w-72 p-0 font-sans font-medium appearance-none focus-visible:ring-2 focus-visible:ring-offset-2 buttonInner" type="button" tabindex="0" data-inner-clickable="true" data-color="primary"><dfn data-align="start" role="presentation" class="icon buttonIcon select-none overflow-hidden inline-block box-content leading-none not-italic text-center rounded" data-clickable="false"><i class="mdi mdi-bell"></i></dfn></button></div></div></div>"`;

exports[`Button should render label only 1`] = `"<div><div class="flex gap-2 items-center p-4"><div class="buttonWrapper relative align-text-top select-none focus-within:z-1 inline-block" data-ref="button"><button role="button" class="button pointer-events-auto cursor-pointer select-none flex flex-nowrap w-full h-full items-center border-0 -outline-offset-1 justify-center max-w-72 p-0 font-sans font-medium appearance-none focus-visible:ring-2 focus-visible:ring-offset-2 buttonInner" type="button" tabindex="0" data-inner-clickable="true" data-color="primary"><label class="buttonLabel truncate" data-colored-icon="false">Click Me</label></button></div></div></div>"`;

exports[`Button should render rounded full-width 1`] = `"<div></div><div></div><div></div><div><div class="flex gap-2 items-center p-4"><div class="buttonWrapper relative align-text-top select-none focus-within:z-1 block w-full" data-ref="button"><button role="button" class="button pointer-events-auto cursor-pointer select-none flex flex-nowrap w-full h-full items-center border-0 -outline-offset-1 justify-center max-w-72 p-0 font-sans font-medium appearance-none focus-visible:ring-2 focus-visible:ring-offset-2 buttonInner max-w-full w-full rounded-full" type="button" tabindex="0" data-inner-clickable="true" data-color="primary"><label class="buttonLabel truncate" data-colored-icon="false">Click Me!</label></button></div></div></div>"`;

exports[`Button should render with badge 1`] = `"<div></div><div></div><div></div><div></div><div></div><div></div><div tabindex="0"></div><div><div class="flex gap-2 items-center p-4"><div class="buttonWrapper relative align-text-top select-none focus-within:z-1 inline-block" data-ref="button"><button role="button" class="button pointer-events-auto cursor-pointer select-none flex flex-nowrap w-full h-full items-center border-0 -outline-offset-1 justify-center max-w-72 p-0 font-sans font-medium appearance-none focus-visible:ring-2 focus-visible:ring-offset-2 buttonInner" type="button" tabindex="0" data-inner-clickable="true" data-color="primary"><label class="buttonLabel truncate" data-colored-icon="false">Click Me</label><div class="contents inline-block"><div data-ref="badge" data-placement="" class="badge inline-block p-px min-w-2 min-h-2 text-center select-none rounded-full leading-none z-5 pointer-events-none relative"><span class="p-1">99</span></div></div></button></div></div></div>"`;
